test/tools/ossfuzz/protoToYul.cpp

```diff
@@ -677,28 +842,167 @@ void ProtoConverter::visit(Block const& _x)
+			m_output << ", ";
+	}
+	m_output << ")";
+	if (_numOutParams > 0)
+	{
+		m_output << " -> ";
+		for (int i = 0; i < _numOutParams; i++)
+		{
+			m_output << "x_" << std::to_string(m_numLiveVars++);
+			m_numVarsPerScope.top()++;
+			if (i < _numOutParams - 1)
+				m_output << ", ";
+			else
+				m_output << "\n";
+		}
+
+	}
+	else
+		m_output << "\n";
+
+	// Body
 	visit(_x.statements());
+	m_numLiveVars -= m_numVarsPerScope.top();
+	m_numVarsPerScope.pop();
+	yulAssert(m_numLiveVars == 0, "Proto fuzzer: Variable stack after function definition is unbalanced.");
+
+	// Manually create a multi assignment using global variables a_0,a_1,a_2,a_3
+	for (int i = 0; i < _numOutParams; i++)
+	{
+		m_output << "a_" << i;
+		if (i < _numOutParams - 1)
+			m_output << ",";
+		else
+			m_output << " := ";
+	}
+
+	m_output << "foo_" << functionTypeToString(_type) << "_" << std::to_string(m_numFunctionSets);
+	m_output << "(";
+	for (int i = 0; i < _numInParams; i++)
+	{
+		m_output << "calldataload(" << std::to_string(i*32) << ")";
+		if (i < _numInParams - 1)
+			m_output << ",";
+	}
+	m_output << ")\n";
+
+	for (int i = 0; i < _numOutParams; i++)
+	{
+		m_output << "sstore(" << std::to_string(i*32) << ", a_" << std::to_string(i) << ")\n";
+	}
+}
+
+void ProtoConverter::visit(FunctionDefinitionNoReturnVal const& _x)
+{
+	int numInParams = _x.num_input_params() % modInputParams;
+	int numOutParams = 0;
+	createFunctionStub(_x, numInParams, numOutParams, NumFunctionReturns::None);
+}
+
+void ProtoConverter::visit(FunctionDefinitionSingleReturnVal const& _x)
+{
+	int numInParams = _x.num_input_params() % modInputParams;
+	int numOutParams = 1;
+	createFunctionStub(_x, numInParams, numOutParams, NumFunctionReturns::Single);
+}
+
+void ProtoConverter::visit(FunctionDefinitionMultiReturnVal const& _x)
+{
+	int numInParams = _x.num_input_params() % modInputParams;
+	// Synthesize at least 2 return parameters and at most (modOutputParams - 1)
+	int numOutParams = std::max((unsigned int)2, _x.num_output_params() % modOutputParams);
+	createFunctionStub(_x, numInParams, numOutParams, NumFunctionReturns::Multiple);
+}
+
+void ProtoConverter::visit(FunctionDefinition const& _x)
+{
+	visit(_x.fd_zero());
+	visit(_x.fd_one());
+	visit(_x.fd_multi());
+	m_numFunctionSets++;
+}
+
+void ProtoConverter::visit(Program const& _x)
+{
+	/* Program template is as follows
+	 *      Four Globals a_0, a_1, a_2, and a_3 to hold up to four function return values
+	 *
+	 *      Repeated function definitions followed by function calls of the respective function
+	 *          Example: function foo(x_0) -> x_1 {}
+	 *                   a_0 := foo(calldataload(0))
+	 *                   sstore(0, a_0)
+	 */
+	m_output << "{\n";
+	// Create globals at the beginning
+	m_output << "let ";
+	for (int i = 0; i < modOutputParams - 1; i++)
+	{
+		m_output << "a_" << i;
+		if (i < modOutputParams - 2)
+			m_output << ", ";
+		else
+			m_output << "\n";
+	}
+	if (_x.funcs_size() > 0)
```

Comment:
Is this required? if `_x.funcs_size() == 0`, doesn't `_x.funcs()` just return an empty list?

Reply:
Yeah, this can be removed.

Reply:
done